src/query_plan/operation.rs

```diff
@@ -0,0 +1,431 @@
+            directives: directives_with_sorted_arguments(&inline_fragment.directives),
+        }
+    }
+}
+
+fn directives_with_sorted_arguments(directives: &DirectiveList) -> DirectiveList {
+    let mut directives = directives.clone();
+    for directive in &mut directives {
+        directive
+            .make_mut()
+            .arguments
+            .sort_by(|a1, a2| a1.name.cmp(&a2.name))
+    }
+    directives
+}
+
+fn flatten_selections(selections: &NormalizedSelectionMap) -> Vec<Selection> {
+    let mut flattened = vec![];
+    for selection in selections.values() {
+        match selection {
+            NormalizedSelection::NormalizedField(normalized_field) => {
+                let selections = flatten_selections(&normalized_field.selection_set.selections);
+                let field = Field {
+                    definition: normalized_field.definition.to_owned(),
+                    alias: normalized_field.alias.to_owned(),
+                    name: normalized_field.name.to_owned(),
+                    arguments: normalized_field.arguments.to_owned(),
+                    directives: normalized_field.directives.to_owned(),
+                    selection_set: SelectionSet {
+                        ty: normalized_field.selection_set.ty.clone(),
+                        selections,
+                    },
+                };
+                flattened.push(Selection::Field(Node::new(field)));
+            }
+            NormalizedSelection::NormalizedInlineFragment(normalized_fragment) => {
+                let selections = flatten_selections(&normalized_fragment.selection_set.selections);
+                let fragment = InlineFragment {
+                    type_condition: normalized_fragment.type_condition.to_owned(),
+                    directives: normalized_fragment.directives.to_owned(),
+                    selection_set: SelectionSet {
+                        ty: normalized_fragment.selection_set.ty.clone(),
+                        selections,
+                    },
+                };
+                flattened.push(Selection::InlineFragment(Node::new(fragment)));
+            }
+        }
+    }
+    flattened
+}
+
+/// Normalizes selection set within specified operation.
+///
+/// This method applies following normalizations
+/// - expands all fragments within an operation
+/// - merge same selections
+/// - removes all introspection fields from top-level selection
+/// - attempts to remove all unnecessary/redundant inline fragments
+pub fn normalize_operation(
+    operation: &mut Operation,
+    _schema: &Schema,
+    fragments: &IndexMap<Name, Node<Fragment>>,
+) {
+    let mut normalized_selection_set =
+        NormalizedSelectionSet::from_selection_set(&operation.selection_set, fragments);
+    // removes top level introspection
+    normalized_selection_set
+        .selections
+        .retain(|key, _| match key {
+            NormalizedSelectionKey::Field { name, .. } => !name.starts_with("__"),
```

Comment:
Shouldn’t this check for `__schema` and `__type` specifically (schema introspection), and let `__typename` through?

Reply:
We only remove those introspection fields from top level selection. Selecting __typename at root should still be handled by router as it useless from query planner perspective.

Reply:
Agreed with @dariuszkuc here, the [existing code](https://github.com/apollographql/federation/blob/7903a296a96948d96f37e2638a6f3e165a413d28/query-planner-js/src/buildPlan.ts#L3427) does the same thing (presumably router already handles this edge case with the existing query planner).